src/file.toit

```diff
@@ -336,3 +352,157 @@ Changes filesystem permissions for the file $name to $permissions.
 */
 chmod name/string permissions/int:
   #primitive.file.chmod
+
+/**
+Copies $source to $target.
+
+If $target is a directory, then takes the basename of $source and appends it to
+  $target.
+
+The location for $target must exist. That is, when copying to `foo/bar`, `foo`
+  must exist.
+
+If $dereference is true, then symbolic links are followed.
+
+If $recursive is true, then directories are copied recursively. If $recursive is
+  false, then $source must be a file.
+*/
+copy --source/string --target/string --dereference/bool=false --recursive/bool=false -> none:
+  // A queue for pending recursive copies.
+  queue := Deque
+  if is-directory target:
+    target = "$target/$(basename_ source)"
+
+  copy_ --source=source --target=target --dereference=dereference --recursive=recursive --queue=queue
+  while not queue.is-empty:
+    next := queue.remove-first
+    new-source := next[0]
+    new-target := next[1]
+    target-stat := stat new-target
+    if not target-stat:
+      throw "Target directory '$new-target' does not exist"
+    target-permissions := target-stat[ST-MODE]
+    // If the The directory was marked as read-only.
+    // Temporarily change the permissions to be able to copy the directory.
+    if system.platform != system.PLATFORM-WINDOWS and target-permissions & 0b010_000_000 != 0b010_000_000:
+      chmod new-target (target-permissions | 0b010_000_000)
+    else if system.platform == system.PLATFORM-WINDOWS and target-permissions & WINDOWS-FILE-ATTRIBUTE-READONLY != 0:
+      // The directory was marked as read-only.
+      // Temporarily change the permissions to be able to copy the directory.
+      target-permissions = target-stat[ST-MODE]
```

Comment:
What does this do? Doesn't `target-permissions` already have the right value?

Reply:
I probable forgot to remove it after I copied it up.
done.